conda/common/compat.py

```diff
@@ -294,3 +296,46 @@ def ensure_utf8_encoding(value):
     except UnicodeEncodeError:
         return value
 
+
+class TemporaryDirectory(object):
+    """Create and return a temporary directory.  This has the same
+    behavior as mkdtemp but can be used as a context manager.  For
+    example:
+
+        with TemporaryDirectory() as tmpdir:
+            ...
+
+    Upon exiting the context, the directory and everything contained
+    in it are removed.
+    """
+
+    # Handle mkdtemp raising an exception
+    name = None
+    _closed = False
+
+    def __init__(self, suffix="", prefix='tmp', dir=None):
+        self.name = mkdtemp(suffix, prefix, dir)
+
+    def __repr__(self):
+        return "<{} {!r}>".format(self.__class__.__name__, self.name)
+
+    def __enter__(self):
+        return self.name
+
+    def cleanup(self, _warn=False, _warnings=_warnings):
+        from ..gateways.disk.delete import rm_rf as _rm_rf
```

Comment:
I think it makes sense to adhere to https://github.com/conda/conda/blob/93db2c14f7b49bce0a00d80a0463470f8a797d50/conda/common/README.rst and not import from outside `conda.common` here to ease later refactoring (e.g., in the advent of a new `libconda`).
Thus, I think it's better to move `TemporaryDirectory` to `conda.gateways.disk` (and use `tempfile.TemporaryDirectory` for Python >=3.2).

Reply:
Thanks, that's good feedback.